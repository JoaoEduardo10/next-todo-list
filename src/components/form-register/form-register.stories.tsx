import { Meta, ComponentStory } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import { FormRegister } from './form-register';
import { GlobalStyles } from '../../styles/globals-styles';
import { theme } from '../../styles/theme';

export default {
  title: 'FormRegister',
  component: FormRegister,
  decorators: [
    (Story) => (
      <ThemeProvider theme={theme}>
        <Story />
        <GlobalStyles />
      </ThemeProvider>
    ),
  ],
} as Meta;

export const TamplateFormRegister: ComponentStory<typeof FormRegister> = () => {
  return <FormRegister />;
};

TamplateFormRegister.parameters = {
  nextRouter: {
    path: '/login',
    asPath: '/login',
    query: {},
  },
};
