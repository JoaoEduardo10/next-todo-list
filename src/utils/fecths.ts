import { json } from 'stream/consumers';

export type TDataProps = {
  email: string;
  password: string;
};
interface IUser {
  name: string;
  email: string;
  password: string;
}
const urlApi = process.env.NEXT_PUBLIC_URL_API as string;

const loginUser = async (data: TDataProps) => {
  const response = await fetch(`${urlApi}/login`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  const token = await response.json();

  return token;
};

const createUser = async (user: IUser) => {
  const { email, name, password } = user;

  const response = await fetch(`${urlApi}/users`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
    },
    body: JSON.stringify({ email, password, name }),
  });

  return {
    ok: response.ok,
  };
};

export { loginUser, createUser };
