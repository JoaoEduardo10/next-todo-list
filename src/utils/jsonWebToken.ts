import jwt from 'jsonwebtoken';

export interface Ijwt {
  email: string;
  name: string;
  id: string | number;
}

export interface IjwtComplete {
  iat: number;
  exp: number;
}

export const compareJwt = (token: string): (Ijwt & IjwtComplete) | null => {
  try {
    const jwtHash = process.env.NEXT_PUBLIC_HASH_JWT as string;

    const varifyToken = jwt.verify(token, jwtHash);

    return varifyToken as Ijwt & IjwtComplete;
  } catch (error: any) {
    console.log(error.message);
    return null;
  }
};
