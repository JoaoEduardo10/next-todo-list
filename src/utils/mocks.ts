import configureStore from 'redux-mock-store';
import { TBoard } from '../types';
import { mockTask } from '../components/task/mock';

const mockStore = configureStore([]);

const mockBoards: TBoard[] = [
  { id: 'board1', boardName: 'Board 1', taskConnect: '123', userId: '2' },
  { id: 'board2', boardName: 'Board 2', taskConnect: '1234', userId: '2' },
];

const store = mockStore({
  boards: {
    allBoards: [
      { id: 'board1', boardName: 'Board 1' },
      { id: 'board2', boardName: 'Board 2' },
    ],
    actualBoard: { id: 'board1', boardName: 'Board 1' },
    actualBoardWithTasks: {
      boardName: 'Board 1',
      id: '1234',
      taskConnect: '1234',
      tasks: mockTask,
      userId: '1234',
    },
  },
});

export { store, mockBoards };
