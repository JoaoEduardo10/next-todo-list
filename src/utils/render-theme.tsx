import { render, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/theme';
import { store as storeActual } from '../app/store';
import { Provider } from 'react-redux';

export const renderTheme = (
  children: React.ReactNode,
  store?: any,
): RenderResult => {
  return render(
    <Provider store={store ?? storeActual}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </Provider>,
  );
};
