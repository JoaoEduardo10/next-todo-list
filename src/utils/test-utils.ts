import { configureStore } from '@reduxjs/toolkit';
import rootReducer from '../../src/app/store';

export const createTestStore = () => {
  const store = configureStore({
    reducer: rootReducer,
    devTools: false,
  });

  return store;
};
