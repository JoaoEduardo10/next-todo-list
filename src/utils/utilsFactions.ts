import { Tasks } from '../components/task/styles';
import { TTasks } from '../types';

const returnedSubTasksConcluded = (task: TTasks) => {
  const allSubTaskConcluded = task.subTasks.filter((total) => {
    return total.concluded == true;
  });

  return allSubTaskConcluded.length;
};

const returnTasksByStatus = (tasks: TTasks[]) => {
  const tasksConcluidedFilter = tasks.filter((task) => {
    return task.status == 'concluded';
  });

  const tasksPendingFilter = tasks.filter((task) => {
    return task.status == 'pending';
  });

  const tasksProgressFilter = tasks.filter((task) => {
    return task.status == 'progress';
  });

  return { tasksConcluidedFilter, tasksPendingFilter, tasksProgressFilter };
};

export { returnedSubTasksConcluded, returnTasksByStatus };
