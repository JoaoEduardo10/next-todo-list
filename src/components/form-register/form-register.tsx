import { createUser } from '../../utils/fecths';
import Link from 'next/link';
import { useEffect, useState } from 'react';
import { Button } from '../button';
import * as S from '../form-login/styles';
import { Inpult } from '../inpult/inpult';
import { MessageError } from '../messageError';
import { useRouter } from 'next/router';
import { Loading } from '../loading';

export const FormRegister = () => {
  const router = useRouter();
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState(false);
  const [primaryRederition, setPrimaryRederition] = useState(false);
  const [loading, setLoading] = useState(false);
  const [messageError, setMessageError] = useState('');

  useEffect(() => {
    let timeError: NodeJS.Timeout;

    if (error) {
      timeError = setTimeout(() => {
        setError(false);
      }, 4000);
    }

    return () => clearTimeout(timeError);
  }, [error]);

  useEffect(() => {
    let timeError: NodeJS.Timeout;

    if (!error) {
      timeError = setTimeout(() => {
        setPrimaryRederition(false);
      }, 5000);
    }

    return () => clearTimeout(timeError);
  }, [error]);

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setPrimaryRederition(false);
    setLoading(true);

    if (!email || !password || !name) {
      setPrimaryRederition(true);
      setError(true);
      setMessageError('Email, name e senha são obrigatórios!');
      setLoading(false);
      return;
    }

    const response = await createUser({ email, name, password });

    if (response.ok == false) {
      setPrimaryRederition(true);
      setError(true);
      setMessageError('Email Ou senha incorretos!');
      setLoading(false);
      return;
    }

    router.push('/login');
    setLoading(false);
  };

  return (
    <S.Form role={'form'} onSubmit={handleSubmit}>
      {loading && <Loading />}
      {primaryRederition && <MessageError error={error} text={messageError} />}
      <S.Heading>Cadastro</S.Heading>
      <Inpult placeholder="Nome" type="text" onChange={setName} value={name} />
      <Inpult
        placeholder="Email"
        type="email"
        onChange={setEmail}
        value={email}
      />
      <Inpult
        placeholder="Senha"
        type="password"
        onChange={setPassword}
        value={password}
      />

      <Button disabled={loading}>Enviar</Button>
      <Link href={'/login'}>SignIn</Link>
    </S.Form>
  );
};
