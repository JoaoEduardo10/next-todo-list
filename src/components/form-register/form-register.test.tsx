import { FormRegister } from './form-register';
import { renderTheme } from '../../utils/render-theme';
import { act, fireEvent, screen, waitFor } from '@testing-library/react';
import fetchMock from 'jest-fetch-mock';

jest.useFakeTimers();
jest.mock('next-auth/react');

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/login',
      pathname: '',
      query: '',
      asPath: '/login',
      push: jest.fn(),
    };
  },
}));

jest.spyOn(require('next/router'), 'useRouter');

describe('<FormRegister />', () => {
  beforeEach(() => {
    fetchMock.enableMocks();
  });

  afterEach(() => {
    fetchMock.resetMocks();
    jest.clearAllMocks();
  });

  it('should render a FormsRegister', () => {
    renderTheme(<FormRegister />);

    const form = screen.getByRole('form');

    expect(form).toBeInTheDocument();
  });

  it('should return an error by not sending the email', async () => {
    renderTheme(<FormRegister />);

    const form = screen.getByRole('form');
    const name = screen.getByPlaceholderText('Nome');
    const password = screen.getByPlaceholderText('Senha');
    const button = screen.getByRole('button');

    expect(form).toBeInTheDocument();

    fireEvent.change(name, { target: { value: 'test' } });
    fireEvent.change(password, { target: { value: 'test123' } });

    fireEvent.click(button);

    const message = await screen.findByLabelText('Message Error');

    expect(message).toBeInTheDocument();
    expect(message).toHaveTextContent('Email, name e senha são obrigatórios!');
    expect(message).toHaveStyle({
      animation: 'showMessage 300ms ease-in-out',
    });

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(message).toHaveStyle({
      animation: 'hidenShowMessageError 1s ease-in-out',
    });
  });

  it('should return an error by not sending the password', async () => {
    renderTheme(<FormRegister />);

    const form = screen.getByRole('form');
    const name = screen.getByPlaceholderText('Nome');
    const email = screen.getByPlaceholderText('Email');
    const button = screen.getByRole('button');

    expect(form).toBeInTheDocument();

    fireEvent.change(name, { target: { value: 'test' } });
    fireEvent.change(email, { target: { value: 'test123@gmail.com' } });

    fireEvent.click(button);

    const message = await screen.findByLabelText('Message Error');

    expect(message).toBeInTheDocument();
    expect(message).toHaveTextContent('Email, name e senha são obrigatórios!');
    expect(message).toHaveStyle({
      animation: 'showMessage 300ms ease-in-out',
    });

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(message).toHaveStyle({
      animation: 'hidenShowMessageError 1s ease-in-out',
    });
  });

  it('should return an error by not sending the name', async () => {
    renderTheme(<FormRegister />);

    const form = screen.getByRole('form');
    const email = screen.getByPlaceholderText('Email');
    const password = screen.getByPlaceholderText('Senha');
    const button = screen.getByRole('button');

    expect(form).toBeInTheDocument();

    fireEvent.change(email, { target: { value: 'test123@gmail.com' } });
    fireEvent.change(password, { target: { value: 'test123' } });

    fireEvent.click(button);

    const message = await screen.findByLabelText('Message Error');

    expect(message).toBeInTheDocument();
    expect(message).toHaveTextContent('Email, name e senha são obrigatórios!');
    expect(message).toHaveStyle({
      animation: 'showMessage 300ms ease-in-out',
    });

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(message).toHaveStyle({
      animation: 'hidenShowMessageError 1s ease-in-out',
    });
  });

  it('should returns an error because the user email already exists', async () => {
    fetchMock.mockResolvedValueOnce({
      ok: false,
    } as any);

    renderTheme(<FormRegister />);

    const form = screen.getByRole('form');
    const email = screen.getByPlaceholderText('Email');
    const password = screen.getByPlaceholderText('Senha');
    const name = screen.getByPlaceholderText('Nome');

    expect(form).toBeInTheDocument();

    fireEvent.change(email, { target: { value: 'test123@gmail.com' } });
    fireEvent.change(password, { target: { value: 'test123' } });
    fireEvent.change(name, { target: { value: 'test' } });

    fireEvent.submit(form);

    const message = await screen.findByLabelText('Message Error');

    expect(message).toBeInTheDocument();
    expect(message).toHaveTextContent('Email Ou senha incorretos!');
    expect(message).toHaveStyle({
      animation: 'showMessage 300ms ease-in-out',
    });

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(message).toHaveStyle({
      animation: 'hidenShowMessageError 1s ease-in-out',
    });
  });

  it('should created a user', async () => {
    fetchMock.mockResolvedValueOnce({
      ok: true,
    } as any);

    renderTheme(<FormRegister />);

    const form = screen.getByRole('form');
    const email = screen.getByPlaceholderText('Email');
    const password = screen.getByPlaceholderText('Senha');
    const name = screen.getByPlaceholderText('Nome');

    expect(form).toBeInTheDocument();

    fireEvent.change(email, { target: { value: 'test123@gmail.com' } });
    fireEvent.change(password, { target: { value: 'test123' } });
    fireEvent.change(name, { target: { value: 'test' } });

    fireEvent.submit(form);

    await waitFor(() => {
      const message = screen.queryByLabelText('Message Error');

      expect(message).not.toBeInTheDocument();
    });

    act(() => {
      jest.advanceTimersByTime(5000);
    });
  });
});
