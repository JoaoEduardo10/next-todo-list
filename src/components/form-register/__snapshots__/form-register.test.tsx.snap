// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormRegister /> should to match snapshot 1`] = `
.c6 {
  margin-top: 3rem;
  width: 16rem;
  padding: 1.6rem;
  background-color: #827FD3;
  border: none;
  border-radius: 1.6rem;
  font-size: 1.8rem;
  line-height: 100%;
  color: #FFFFFF;
  cursor: pointer;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
}

.c6:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c6:disabled {
  background-color: #F4F7FD;
  cursor: not-allowed;
  color: #000;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c4 {
  width: 80%;
  padding: 1.6rem;
  border: none;
  background-color: #FFFFFF;
  outline: none;
  line-height: 1.2;
  font-size: 1.7rem;
}

.c3 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1.6rem;
  border: none;
  background-color: #FFFFFF;
  gap: 0.5rem;
  box-shadow: 1rem 1rem 1.6rem #000;
  border-radius: 0.9rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 1.2;
  -webkit-transition: all 300ms ease-in;
  transition: all 300ms ease-in;
}

.c3 svg {
  font-size: 2.6rem;
  cursor: pointer;
}

.c0 {
  position: relative;
  background-color: #F4F7FD;
  width: 40%;
  height: 95%;
  padding: 4.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 4rem;
  border-radius: 0.5rem;
  box-shadow: 0.1rem 0.1rem 0.2rem #FFFFFF;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 a {
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 1.6rem;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
}

.c0 a:hover {
  color: #827FD3;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 {
  margin-bottom: 6rem;
  color: #827FD3;
  font-family: Arial,Helvetica,sans-serif;
  font-size: 3.8rem;
}

@media (max-width:780px) {
  .c0 {
    width: 90%;
    padding: 1.6rem;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }

  .c0 .c2 {
    width: 100%;
  }
}

@media (max-width:950px) and (min-width:781px) {
  .c0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    width: 80%;
    padding: 0 4.8rem;
  }

  .c0 .c2 {
    height: 10rem;
    margin-bottom: 2.4rem;
  }

  .c0 .c2 input {
    font-size: 3.6rem;
  }

  .c0 .c2 svg {
    font-size: 4.4rem;
  }

  .c0 .c5 {
    width: 50%;
    height: 10rem;
    font-size: 4rem;
  }

  .c0 a {
    font-size: 3rem;
  }
}

@media (max-width:1500px) and (min-width:951px) {
  .c0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }

  .c0 .c2:nth-child(3) {
    margin-bottom: 0.9;
  }

  .c0 .c2 {
    height: 7rem;
    width: 100%;
  }

  .c0 .c2 input {
    font-size: 2.6rem;
  }

  .c0 .c2 svg {
    font-size: 3.4rem;
  }

  .c0 .c5 {
    width: 90%;
    height: 6rem;
    font-size: 3rem;
  }

  .c0 a {
    font-size: 2.4rem;
  }
}

@media (max-width:780px) {
  .c1 {
    margin-bottom: 4rem;
  }
}

@media (max-width:950px) and (min-width:781px) {
  .c1 {
    font-size: 8rem;
  }
}

@media (max-width:1500px) and (min-width:951px) {
  .c1 {
    font-size: 4rem;
    margin-bottom: 1.6rem;
  }
}

<form
  class="c0"
  role="form"
>
  <h1
    class="c1"
  >
    Cadastro
  </h1>
  <label
    aria-label="Label conteiner"
    class="c2 c3"
    for="text"
  >
    <input
      class="c4"
      id="text"
      placeholder="Nome"
      type="text"
      value=""
    />
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M18.41 5.8L17.2 4.59c-.78-.78-2.05-.78-2.83 0l-2.68 2.68L3 15.96V20h4.04l8.74-8.74 2.63-2.63c.79-.78.79-2.05 0-2.83zM6.21 18H5v-1.21l8.66-8.66 1.21 1.21L6.21 18zM11 20l4-4h6v4H11z"
      />
    </svg>
  </label>
  <label
    aria-label="Label conteiner"
    class="c2 c3"
    for="email"
  >
    <input
      class="c4"
      id="email"
      placeholder="Email"
      type="email"
      value=""
    />
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
      />
    </svg>
  </label>
  <label
    aria-label="Label conteiner"
    class="c2 c3"
    for="password"
  >
    <input
      class="c4"
      id="password"
      placeholder="Senha"
      type="password"
      value=""
    />
    <svg
      aria-label="Password close"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath>
          <path
            clip-rule="evenodd"
            d="M124-288l388-672 388 672H124z"
            fill="none"
          />
        </clippath>
      </defs>
      <path
        d="M508 624a112 112 0 0 0 112-112c0-3.28-.15-6.53-.43-9.74L498.26 623.57c3.21.28 6.45.43 9.74.43zm370.72-458.44L836 122.88a8 8 0 0 0-11.31 0L715.37 232.23Q624.91 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 0 0 0 51.5q56.7 119.43 136.55 191.45L112.56 835a8 8 0 0 0 0 11.31L155.25 889a8 8 0 0 0 11.31 0l712.16-712.12a8 8 0 0 0 0-11.32zM332 512a176 176 0 0 1 258.88-155.28l-48.62 48.62a112.08 112.08 0 0 0-140.92 140.92l-48.62 48.62A175.09 175.09 0 0 1 332 512z"
      />
      <path
        d="M942.2 486.2Q889.4 375 816.51 304.85L672.37 449A176.08 176.08 0 0 1 445 676.37L322.74 798.63Q407.82 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 0 0 0-51.5z"
      />
    </svg>
  </label>
  <button
    class="c5 c6"
  >
    Enviar
  </button>
  <a
    href="/login"
  >
    SignIn
  </a>
</form>
`;
